Rx js

Maccolm opened the pull request.

Maccolm committed:
added ngOnDestroy on each page where have subscription

Maccolm committed:
realized interaction with watch and favorite lists through the behavior Subject;
- added types to functions from <any> to <Movie>...

renamed "watchlist" component to "watchList"

Maccolm committed:
changed styling, hovers, clicks, for better looking

Maccolm committed:
get account ID
added unsubscribe "clearObservable"
renamed observable "$" for other developers

Maccolm committed:
favorite list taking from API

Maccolm committed:
trying to add delete function as observable

Maccolm committed:
made setting with adding to favorite list through api
- delete from favorite list
- it appears on all pages

Maccolm committed:
fixed methods with watchList through API
- add to watchList
- removeFrom watchList
- load from watchlist and add it to subject after reload a page
- synchronize watch list on different pages after adding to watchList and after reloading a page.
- synchronize watch list on all pages after deleting a movie from watchList

inga-lekhman commented on src/app/pages/now-playing-movie-page/now-playing-movie-page.component.ts:
```diff
@@ -1,25 +1,30 @@
   imports: [MovieCardComponent, MovieHeaderComponent],
 })
-export class MovieNowPlayingPageComponent implements OnInit {
+export class MovieNowPlayingPageComponent extends ClearObservable implements OnInit {
```
Якщо ви створили цей клас, то наслідуйте його всюди, де потрібно) ClearObservable

inga-lekhman commented on src/app/app.component.ts:
```diff
@@ -1,29 +1,63 @@
+        console.error('Authentication failed:', error);
+      }
+    );
+//1) тут я отримую список улюблених та watchlist фільмів та сетаю в сабжект, це працює після оновлення сторінки
```
А навіщо ви це робите? Вам це не потрібно. Вам потрібно робити this.movieService.getFavoriteMovies() при переход іна сторінку **watchList**

Те саме стосується favorites.

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -1,7 +1,7 @@
 import { Injectable } from '@angular/core';
```
Старайтесь форматувати файл при збереженні (забирати лишні пробіли, пусті рядки, ставити крапку з комою в кінці і т.д.). Це можна налаштувати в IDE.
Тоді ваш файл буде чистим, читабельним і легко підтримуваним.

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -10,12 +10,35 @@ export class MovieService {
+	private accountId: number | null = this.getAccountId();
+	private sessionId: string = this.getSessionId();
+	
+	private apiUrl = 'https://api.themoviedb.org/3/movie';
```
Рекмондую такі штуки зберігати в **env** файлах. На майбутнє )

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -29,45 +52,105 @@ export class MovieService {
+	  favorite: false
+	};
+	return this.httpClient.post<Movie>(url, body).pipe(
+		response => response,
```
що цей рядок робить? Здається, він лишній

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -29,45 +52,105 @@ export class MovieService {
+		const url = `${this.baseUrl}/account/${this.accountId}/watchlist/movies${this.apiKey}&session_id=${this.sessionId}`;
+		return this.httpClient.get<{ results: Movie[] }>(url).pipe(
+			map(response => response.results),
+			tap(movies => this.watchListSubject$.next(movies)),
```
Цей сет вам потрібно робити в компоненті, де ви вмкликаєте цей метод.
Тут ви використовуєте **tap** оператор, і він трохи збиватиме вас у майбетньому. Цей оператор зазвичай використовується для дебаггінгу.

А тут у вас кол до сервера, тому краще такого роду сайд ефекти забрати і сетити через  виклик **setToWatchList** в компоненті, де ви викликаєте **getWatchList**

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -29,45 +52,105 @@ export class MovieService {
+	 };
+	 this.setToWatchListSubject$(movie)
+	 return this.httpClient.post<Movie>(url, body).pipe(
+		map(response => response),
```
цей map не потрібен тут.

inga-lekhman commented on src/app/pages/favorite-list-movie-page/favorite-list-movie-page.component.ts:
```diff
@@ -9,16 +12,32 @@ import { MovieService } from '../../services/movie.service';
+	deleteFromFavorites(movie: Movie) {
+		this.movieService.removeFromFavoriteMovies(movie).pipe(takeUntil(this.destroy$)).subscribe(response => {
+			console.log(response);
+			if(response) {
```
це сміло можете видаляти. Вам не потрібно після видалення оновлювати список favorites.
Він у вас автоматично оновиться завдяки рядку 25: 
this.movieService.getFavoriteMovies().pipe(takeUntil(this.destroy$)).subscribe(movies => {
			this.favoriteMovies = movies;
		},
		
Це у вас підписка на зміни в списку favorites. Однак в самому сервісі у вас нема видалення з цього масиву, тому додайте його. Там буде комент.

inga-lekhman commented on src/app/pages/movie-watchlist-page/movie-watchlist-page.component.ts:
```diff
@@ -1,25 +1,43 @@
+  deleteFromWatchList(movie: Movie) {
+    this.movieService.deleteFromWatchList(movie).pipe(takeUntil(this.destroy$)).subscribe(response => {
+		console.log(response);
+		if (response) {
```
Тут те саме, дивіться комент вище.

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -29,45 +52,105 @@ export class MovieService {
   }
-  getWatchList() {
-    return this.watchList;
+  setToFavoriteMovies(movie: Movie) {
```
Я бачу по додатку, що що сет і ремов для favorites and watchList вертає 401 помилку. 
![image](https://github.com/user-attachments/assets/b02474d7-0324-4804-9d60-6c22ad2aa58a)

Тому у вас щось не правильно для цих АПІ реквестів задається. 

І ще є такий м омент стосовно цієї функціональності: зазвичай, якщо для того, шоб додати до **favorites/watchList** вивикорситовуєте сервер, тоді на сервері має оброблятися додавання і видалення, а не на UI. 

UI має просто отримати оновлені дані з сервера після того, як відповідний елемент додався-видаляився.

Наразі у вас це все робиться в цьому файлі, що створює незручності в подальшій роботі з цим сервісом і спричинить багато непорозумінь та помилок в майбутньому.

inga-lekhman replied:
Це ви пофіксили наскільки я бачу

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -29,45 +52,105 @@ export class MovieService {
+	  media_id: movie.id,
+	  favorite: true
+	};
+	this.setToFavoriteMoviesSubject(movie)
```
якщо ви тут додаєте елемент, то вам потрібно у відповідній функції так само його видаляти: тут removeFromFavoriteMovies.

Тоді вам не потрібно буде писати _subscribe hell_ у відповідних компонентах, якраз те, що ви маєте наразі.

inga-lekhman reviewed (commented):
Отже, у вас досить заплатуна логіка з **favorites and watch list.** 
Відповідно вона у вас не працює належним чином на фронтенді. 

Перегляньте те, що я вам написала і перевірте ще раз все детально, видаліть лишнє і я впевнена, все запрацює )

Maccolm committed:
changed auth service.
- added environments
- changed favorites and watchlist logic

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -1,73 +1,149 @@
+      media_id: movie.id,
+      favorite: true,
+    };
+    this.setToFavoriteMoviesSubject(movie);
```
У вас наразі трішки заплутана логіка із оновленням **favorites and watch list.**

**По-перше**, ви при ініціалізації додатку, завантажуєте всі favorites and watch list і потім ще сетите їх в циклі за допомогою методу: **setToFavoriteMovies**, тобто тут ви відправляєте запит на сервер для фільмів, які і так уже є в улюблених - це лишні кроки. 
**Що робити? -** На етапі ініціалізації додатку, вам вистачить просто завантажити **favorites and watch list,** і дійсно це можна зробити в **app.component.** Після того, як ви їх завантажите, вам потрібно засетити їх в **favoriteMoviesSubject$**, це дозволить підписникам отримати свіжі дані. В ролі підписників у вас компоненти **movie-card** наскільки я бачу. Відповідно, всередині цих компонентів ви можете отримати стан: чи є даний фільм у favorites/watchList.

**По-друге,** вам нема сенсу при додаванні/видаленні фільмів до/з favorite/watchList викликати метод this.setToFavoriteMoviesSubject(movie), оскільки вам не потрібно, щоб компонента movie-card, в якій ця дія(додавання/видалення) відбувається, ще й оновлювалася, коли ви клакнете додати чи видалити. 

**Що робити?** - В компоненті **movie-card** у вас є змінні **isInFavorite** та **isInWatchList**, які відповідатимуть за цей стан фільму. При ініціалізації ви їх сетите, а при кліку, просто оновлюєте.
метод setToFavoriteMovies буде сетити цей фільм на сервері. Тоді при відкритті сторінки favorites/watchList ви будете завантажувати дані з сервера, в яких вже будуть всі необхідні елементи, і тим самим оновите **favoriteMoviesSubject$**

**І по-третє:** у випадку, якщо у вас спільні фільми в різних категоріях, у вас тоді фільми, які ви, для прикладу, додали на сторінці http://localhost:4200/popular до улюблених, на іншій сторінці(наприклад http://localhost:4200/top-rate) не будуть позначені як додані до улюблених.

**Що робити? -** те. що ви робите для **app.component,** робите просто для кожної **page**-компоненти. В **Angular** це можна зробити в одному місці при зміні роута (в **резолвері**), але поки можете продублювати. Або зробіть резолвер, він додається до кожного роута.
Тобто перед тим, як перейти на роут, ви отримаєте дані з сервера для **favorites/watchList.**

На перший погляд складно, але якщо ви вжде вирігшили зберігати вибрані фільми після перезавантаження сторінки, то потрібно зробити правильно.

Якщо що, пишіть )

inga-lekhman commented on src/app/app.component.ts:
```diff
@@ -1,29 +1,48 @@
 
-	constructor() {}
+  ngOnInit(): void {
+	//коли ціх субскрайбів немає, то при оновленні сторінки помітки in favorite, watchlist зникають. Тільки з ними вони залишаються на місці.
```
Дивіться розширений комент в **movie.service.ts**

inga-lekhman commented on src/app/components/card-movie/movie-card.component.ts:
```diff
@@ -30,23 +31,48 @@ export class MovieCardComponent implements OnInit {
-    this.isInWatchList = this.movieService.isInWatchList(this.movie);
+	 this.subscriptions.add(
+		this.movieService.favoriteMovies$.subscribe(favorites => {
+			this.isInFavorite = this.movieService.isInFavoriteList(this.movie);
```
На мою думку, цей метод в сервісі зайвий - краще робіть цю перевірку тут - це один рядок. А там видаліть. 

Це буде більш зрозуміло, читабельно та підтримувано. Інакше - ви в цьому компоненті звертайєтесь до сервіса, отримуєте дані і знову звертаєтесь до сервіса (навіщо, це можна було відразу в сервісі зробити і повернути те, що треба).

inga-lekhman commented on src/app/pages/description-movie-page/description-movie-page.component.ts:
```diff
@@ -26,32 +28,36 @@ export class MovieDescriptionComponent {
+	this.subscriptions.add(paramsSub);
   }
   
   addToFavorites() {
```
Я не бачу у вас на сторінці фільму кнопок додавання до **favprites/watchList.** Мабуть ці методи зайві.

inga-lekhman replied:
Цього нема досі

Maccolm replied:
Зроблено

inga-lekhman commented on src/app/pages/description-movie-page/description-movie-page.component.ts:
```diff
@@ -26,32 +28,36 @@ export class MovieDescriptionComponent {
+	this.subscriptions.add(paramsSub);
   }
   
   addToFavorites() {
```
Я не бачу у вас на сторінці фільму кнопок додавання до **favprites/watchList.** Мабуть ці методи зайві.

inga-lekhman commented on src/app/pages/favorite-list-movie-page/favorite-list-movie-page.component.ts:
```diff
@@ -9,16 +12,26 @@ import { MovieService } from '../../services/movie.service';
+	//в мене автоматично не змінювалися фільми, тільки при оновленні сторінки. Вже не знаю як зробити, щоб воно прибирало видалені фільми окрім цього
+	deleteFromFavorites(movie: Movie) {
+		this.movieService.removeFromFavoriteMovies(movie).pipe(takeUntil(this.destroy$)).subscribe(() => {
+			this.loadFavoriteMovies();
```
Ви тут просто змінюйте **this.favoriteMovies**. Цього достатньо. Тоді це не потрібно буде викликати: **this.loadFavoriteMovies();**

inga-lekhman commented on src/app/pages/movie-watchlist-page/movie-watchlist-page.component.ts:
```diff
@@ -1,25 +1,37 @@
+ 	}
+  deleteFromWatchList(movie: Movie) {
+    this.movieService.deleteFromWatchList(movie).pipe(takeUntil(this.destroy$)).subscribe(() => {
+		this.loadWatchList();
```
Тут те саме

inga-lekhman commented on src/app/pages/popular-movie-page/popular-movie-page.component.ts:
```diff
@@ -10,15 +12,23 @@ import { MovieService } from '../../services/movie.service';
 })
-export class MoviePopularPageComponent implements OnInit {
-	popularMovies: any = [];
+export class MoviePopularPageComponent implements OnInit, OnDestroy {
```
Старайтесь форматувати файл при збереженні (забирати лишні пробіли, пусті рядки, ставити крапку з комою в кінці і т.д.). Це можна налаштувати в IDE.
Тоді ваш файл буде чистим, читабельним і легко підтримуваним.

inga-lekhman reviewed (approved):
Чудова робота! Додаток мені дуже подобається і дуже класно, що після завантаження теж все зберігається. 

Я вам написала як змінити основну логіку в movie.service. Якщо ви це зробите, ви побачите, наскільки все інше просто і зручно.

Спробуйте!

Maccolm committed:
preparing resolver for routes

Maccolm committed:
added resolver to pages
changed logic of fv and watch lists
changed card component and function isInFavorite - watchList

Maccolm committed:
fixed problem with auth.service
now it's getting account and sessionId and wokrs

inga-lekhman commented on src/app/app.component.ts:
```diff
@@ -1,29 +1,65 @@
+	this.authService.authenticateAndGetAccountId().subscribe(
+      data => {
+			const { accountId, sessionId } = data
+        this.movieService.setAccountId(accountId);
```
В загальному, краще не сетати акаунт id в інший сервіс - краще з іншого сервісу цей акаунт id просто брати. Типу збережіть його в authService,  а в іншому сервісі (movieService) просто отримайте його, типу **authService.getAccountId()**

Логічно ж зберігати акаунт id в **authService**. Те саме стосується сесії.

Maccolm replied:
Зроблено

inga-lekhman commented on src/app/app.component.ts:
```diff
@@ -1,29 +1,65 @@
+		this.movieService.getFavoriteMovies().pipe(takeUntil(this.destroy$)).subscribe(movies => {
+			const favoriteMovies = movies
+			favoriteMovies.forEach(movie => {
+				this.movieService.setToFavoriteMoviesSubject(movie);
```
Так добре, можливо я минулого разу подумала, що ви відправляєте запит на сервер для кожного фільму. 
так ок.

inga-lekhman commented on src/app/components/card-movie/movie-card.component.ts:
```diff
@@ -30,23 +32,37 @@ export class MovieCardComponent implements OnInit {
   public IMAGINE_PATH: string = 'https://image.tmdb.org/t/p/w500/';
 
-  constructor(private movieService: MovieService, private router: Router) {}
+  private subscriptions: Subscription = new Subscription;
```
Ви тут забули () - new Subscription();

Maccolm replied:
дякую

inga-lekhman commented on src/app/components/card-movie/movie-card.component.ts:
```diff
@@ -30,23 +32,37 @@ export class MovieCardComponent implements OnInit {
+			favorites.find(m => m.id === this.movie.id) ? this.isInFavorite = true : this.isInFavorite = false;
+	});
+	this.movieService.watchList$.pipe(takeUntil(this.destroy$)).subscribe(watchList =>{
+		watchList.find(m => m.id === this.movie.id) ? this.isInWatchList = true : this.isInWatchList = false;
```
Ви можете це спростити:
`watchList.find(m => m.id === this.movie.id) ? this.isInWatchList = true : this.isInWatchList = false;`

та написати 
`this.isInWatchList = Boolean(watchList.find(m => m.id === this.movie.id));`

inga-lekhman commented on src/app/components/card-movie/movie-card.component.ts:
```diff
@@ -30,23 +32,37 @@ export class MovieCardComponent implements OnInit {
+		response => {
+			console.log('added to fv', response);
+		});
     this.isInFavorite = true;
```
Це краще додати в subscribe блок, оскільки ви маєте сетати це тільки тоді, коли сервер повернув Success(200 response). Інакше може бути ситцуація, коли сервер не засетит favorite, а ви тут засетили.

Те саме стосується і watchList.

Maccolm replied:
Дякую, логічно. Так зроблю і з додаванням та видалянням фільмів

inga-lekhman commented on src/app/directives/clearObservable.ts:
```diff
@@ -0,0 +1,12 @@
+    destroy$: Subject<boolean> = new Subject();
+    
+    ngOnDestroy() : void {
+        this.destroy$.next(true);
```
В цьому випадку true як параметер не потрібен - не має сенсу. Тип може бути **void**.

Maccolm replied:
в мене чомусь так помилку видає

inga-lekhman commented on src/app/pages/favorite-list-movie-page/favorite-list-movie-page.component.ts:
```diff
@@ -9,16 +13,30 @@ import { MovieService } from '../../services/movie.service';
 	ngOnInit(): void {
-		this.favoriteMovies = this.movieService.getFavoriteMovies();
+		this.route.data.pipe(takeUntil(this.destroy$)).subscribe(data => {
+			this.favoriteMovies = data['favoriteMovies'];
```
Надавайте перевагу доступу по . замість []. Це захистить вас від помилок при перейменуванні полів.

inga-lekhman commented on src/app/routerGuards/favoriteMovies.resolver.ts:
```diff
@@ -0,0 +1,22 @@
+ private fv: Movie[] = [];
+  constructor(private movieService: MovieService) {}
+
+  resolve(route: ActivatedRouteSnapshot): Observable<Movie[]> {
```
Якщо ви зайдете безпосередньо на сторінку favorites, ви зможете побачити що у вас двічі відправляється реквест для отримання фільмів:
![image](https://github.com/user-attachments/assets/45412e09-8422-44d3-a197-79071f59b47d)

Це через те, що ви це робите в app.component і в резолвері. Вам достатньо це робити в одному місці. App.component завжди буде апершим запускати код, незалежно від того, яку сторінку ви відкриєте.

inga-lekhman replied:
Також як відповідь на ваше питання стосовно резолверів:
1) якщо ви викликаєте отримання favorites and watchList в app.component, то вам банально не потрібні ці резолвери, оскільки дані уже будуть завантажені до того, як ви перейдете на відповідні сторінки.

2) Ви можете просто в цих резолверах повертати дані з того сабджекту - там же ж завжди будуть актуальні дані, ви ж його завжди оновлюєте.

3) На випадок якщо раптом чомусь даних нема(null), то можете викликати this.movieService.getFavoriteMovies()

inga-lekhman commented on src/app/pages/favorite-list-movie-page/favorite-list-movie-page.component.ts:
```diff
@@ -9,16 +13,30 @@ import { MovieService } from '../../services/movie.service';
 	}
+	deleteFromFavorites(movie: Movie) {
+		this.movieService.removeFromFavoriteMovies(movie).pipe(takeUntil(this.destroy$)).subscribe(() => {
+			this.updateFavoriteMovies();
```
Я розумію, чому ви так зробили, але ж чи не буде простіше, якщо ви тут просто видалите фільм з локального масиву this.favoriteMovies and subject в сервісі ? 

тоді фільм відразу пропаде, а при наступному завантаженні сервер поверне правильний результат.
тоді вам зайвий раз не треба буде викликати this.movieService.getFavoriteMovies()

inga-lekhman commented on src/app/routerGuards/watchList.resolver.ts:
```diff
@@ -0,0 +1,22 @@
+  resolve(route: ActivatedRouteSnapshot): Observable<Movie[]> {
+    return this.movieService.getWatchList().pipe(
+      tap((movies) => {
+        this.watchList = movies;
```
Це лишня змінна

inga-lekhman commented on src/app/pages/movie-watchlist-page/movie-watchlist-page.component.ts:
```diff
@@ -1,25 +1,42 @@
-  deleteFromWatchList(movie: any){
-	this.movieService.deleteFromWatchList(movie)
+	this.route.data.pipe(takeUntil(this.destroy$)).subscribe(data =>{
+		this.watchList = data['watchList'];
```
Надавайте перевагу доступу по . замість []. Це захистить вас від помилок при перейменуванні полів.

inga-lekhman commented on src/app/pages/movie-watchlist-page/movie-watchlist-page.component.ts:
```diff
@@ -1,25 +1,42 @@
+ 	}
+  deleteFromWatchList(movie: Movie) {
+    this.movieService.deleteFromWatchList(movie).pipe(takeUntil(this.destroy$)).subscribe(() => {
+		this.updateWatchList();
```
Тут те саме як і при favorites.

inga-lekhman commented on src/app/pages/upcoming-movie-page/upcoming-movie-page.component.ts:
```diff
@@ -11,13 +11,19 @@ import { MovieService } from '../../services/movie.service';
+		this.upcoming = movies.results;
+	});
+  }
+  ngOnDestroy(): void {
```
Наслідуйте тут ваш спеціальний клас

Maccolm replied:
Зробив, для всіх pages

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -1,73 +1,197 @@
+	private apiUrl = environment.apiUrl;
+	private baseUrl = environment.baseUrl;
+	
+	private accountIdSubject$ = new BehaviorSubject<number | null>(this.getAccountId());
```
Якщо це movie service, то відповідно він має працювати тільки з фільмами. Усю іншу інфу ви отримуєте з відповідних сервісів: **single responsibility** from SOLID принципів.

Maccolm replied:
Дякую, переніс

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -1,73 +1,197 @@
+		window.localStorage.setItem('account_id', accountId.toString());
+		this.accountIdSubject$.next(accountId);
+	}
+	getAccountId() {
```
Ці методи можете перенести у auth service.

Maccolm replied:
Зроблено!

inga-lekhman commented on src/app/services/movie.service.ts:
```diff
@@ -1,73 +1,197 @@
+		);
+	}
+	setToFavoriteMovies(movie: Movie) {
+		return combineLatest([this.accountId$, this.sessionId$]).pipe(
```
Щоб не створювати таку складну контсрукцію викликів, ви можете створити резолвер, який буде отримувати акаунт айді та сесію, або просто сесію(для головного роута). 

осакільки акаунт айді вам потрібен тільки при переході на роути **favorites** and **watchList**, то також для них можете створити резолвер, який буде робити відповідний виклик на сервер.

Тоді ви уникнути такого роду combineLatest пекла)

Maccolm replied:
Всю цю конструкцію прибрав, зачистив кодб combineLatest пекла більше нема))

Тепер акаунт айді та сесія сетиться у відповідний сервіс, але спочатку в мене при оновленні сторінки повертало null. То я ініціалізую акаунт айді та сесію в момент, коли викликаю favoriteMovies, а favoriteMovies я перший раз викликаю в appComponent і при умові, якщо отримую з серверу акаунт айді та сесію. 

резолвери я прибрав, бо не бачу в них більше необхідності. Можливо на майбутнє, коли сессія буде expired, то воно редіректнить на логін сторінку

inga-lekhman reviewed (approved):
Мені подобається, що ви переробили і думаєте про те, щоб зробити все правильно. 
Але повірте - ви будете розуміти як зробити це правильно з кожним наступним разом. І це ПРАВИЛЬНО - щоращу буде іншим )

Тому почитайте мої коменти, я думаю, там уже все точно має бути зрозумілим, але якщо ні, то пишіть. 

Пропоную удосконалювати код, роблячи це уже в наступних домашках, щоб не відстати )

Maccolm committed:
changed logic of updating favorite and watchList,
took away resolver,
using more subjects for taking data
updating subject after updating data

Maccolm committed:
transfer logic to their services to follow SOLID method.
made authentication more simple
solved problem with error 401, after first app loading

Maccolm committed:
deleted update Favorite and watch lists functions
